spot_driver_cpp/src/interfaces/spot_interface.cpp

```diff
@@ -0,0 +1,394 @@
+
+  const auto authenticate_result = robot_->Authenticate(username, password);
+  if(!authenticate_result)
+  {
+    return tl::make_unexpected("Authentication with provided username and password did not succeed.");
+  }
+
+  // Start time synchronization between the robot and the client system.
+  // This must be done only after a successful authentication.
+  const auto start_time_sync_response = robot_->StartTimeSync();
+  if (!start_time_sync_response)
+  {
+    return tl::make_unexpected("Failed to start time synchronization.");
+  }
+
+  const auto get_time_sync_thread_response = robot_->GetTimeSyncThread();
+  if (!get_time_sync_thread_response)
+  {
+    return tl::make_unexpected("Failed to get the time synchronization thread.");
+  }
+  time_sync_thread_ = get_time_sync_thread_response.response;
+
+  const auto image_client_result =
+        robot_->EnsureServiceClient<::bosdyn::client::ImageClient>(
+            ::bosdyn::client::ImageClient::GetDefaultServiceName());
+  if (!image_client_result.status)
+  {
+    return tl::make_unexpected("Failed to initialize the Spot SDK image client.");
+  }
+
+  image_client_.reset(std::move(image_client_result.response));
+
+  return {};
+}
+
+tl::expected<bool, std::string> SpotInterface::hasArm() const
+{
+  // Determine if Spot has an arm by checking if the client for gripper camera parameters exists, since Spots without arms do not have this client.
+  const auto list_result = robot_->ListServices();
+  if (!list_result.status)
+  {
+    return tl::make_unexpected("Failed to retrieve list of Spot services.");
+  }
+
+  const auto& services = list_result.response;
+
+  return std::find_if(services.cbegin(),services.cend(), [](const ::bosdyn::api::ServiceEntry& entry){ return entry.name() == ::bosdyn::client::GripperCameraParamClient::GetDefaultServiceName(); } ) != services.cend();
+}
+
+tl::expected<GetImagesResult, std::string> SpotInterface::getImages(::bosdyn::api::GetImageRequest request)
+{
+  std::shared_future<::bosdyn::client::GetImageResultType> get_image_result_future = image_client_->GetImageAsync(request);
+
+  ::bosdyn::client::GetImageResultType get_image_result = get_image_result_future.get();
+  if (!get_image_result.status) {
+      return tl::make_unexpected("Failed to get images: " + get_image_result.status.DebugString());
+  }
+
+  const auto clock_skew_result = getClockSkew();
+  if (!clock_skew_result)
+  {
+    return tl::make_unexpected("Failed to get latest clock skew: " + clock_skew_result.error());
+  }
+
+  GetImagesResult out;
+  for (const auto& image_response : get_image_result.response.image_responses())
+  {      
+      const auto& image = image_response.shot().image();
+      auto data = image.data();
+
+      const auto image_msg = toImageMsg(image_response.shot(), robot_name_, clock_skew_result.value());
+      if (!image_msg)
+      {
+        std::cerr << "Failed to convert SDK image response to ROS Image message: " << image_msg.error() << std::endl;
```

Comment:
Do we want to pass in a logger instead of going to cerr?  This would allow us to pass in the ROS logger and have these messages go in logs and stuff without a ROS dependency.

Reply:
I can make these return as the "unexpected" result from this function, which will cause the messages to be logged to the user in SpotImagePublisher. I was sort of on the fence about doing that, but since the reasons these failure can happen  a )are not spurious/intermittent, and b) are indicative of either an implementation bug or something that requires corrective action by the user, I think it's more useful to stop publishing images and return the error immediately.